flask/app/manage_oidc.py

```diff
@@ -0,0 +1,284 @@
+        "redirectUris": [
+            app_url + "/*",
+            "http://localhost:5000/*",
+            "http://localhost:3000/*",
+        ],
+        "webOrigins": ["+"],
+    }
+    response = requests.post(
+        url, json=new_client, headers={"Authorization": f"Bearer {access_token}"}
+    )
+    if response.ok:
+        print("Stager client added successfully.")
+        return True
+    elif response.status_code == 409:
+        print("Stager client already exists. Skipping...")
+        return True
+    app.logger.error(
+        f"ERROR: Failed to add Stager client: {response.status_code} - {response.text}"
+    )
+    return False
+
+
+def add_keycloak_user(access_token: str, user: User, password: str = None):
+    """
+    Using the given admin access token, create a new user based on the
+    given user object.
+
+    Update the provided user with issuer, subject fields and return True if successful.
+    Return False if unsuccessful.
+    """
+    print(f"Adding user {user.username} to Keycloak...")
+    url = keycloak_host + "/auth/admin/realms/ccm/users"
+    # If a password is provided, then we create a user with that password
+    # Otherwise, we specify that this user must set a password themselves
+    # and they login the first time with their username
+    if password:
+        user_credentials = {"type": "password", "value": password, "temporary": False}
+    else:
+        user_credentials = {
+            "type": "password",
+            "value": user.username,
+            "temporary": True,
+        }
+
+    # https://www.keycloak.org/docs-api/12.0/rest-api/index.html#_userrepresentation
+    new_user = {
+        "email": user.email,
+        "emailVerified": True,
+        "enabled": not user.deactivated,
+        "username": user.username,
+        "credentials": [user_credentials],
+    }
+    post_response = requests.post(
+        url, json=new_user, headers={"Authorization": f"Bearer {access_token}"}
+    )
+    if post_response.ok:
+        print(f"User {user.username} added successfully.")
+        # We need to fetch the ID that keycloak assigned to this user so we can store it
+        params = {
+            "username": user.username,
+            "exact": True,
+        }
+        get_response = requests.get(
+            url, params=params, headers={"Authorization": f"Bearer {access_token}"}
+        )
+        if get_response.ok:
+            # Returns array of UserRepresentations
+            found_users = get_response.json()
+            if len(found_users) != 1:
+                app.logger.error(f"ERROR: Added user {user.username} not found ???")
+                return False
+
+            added_user = found_users[0]
+            print(f"Found user in Keycloak: {added_user}")
+            endpoints = fetch_admin_well_known("ccm")
+            issuer = endpoints["issuer"]
+            subject = added_user["id"]
+            user.set_oidc_fields(issuer, subject)
```

Comment:
I needed to commit this via a db session for the changes to take.

Reply:
The issue here is that `add_keycloak_user` is used in seed functions in `manage.py`, which handles committing the db session. I can get around this by adding click commands for these actions that handle the db session logic separately.